spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java

```diff
@@ -24,13 +24,17 @@
 import java.nio.ByteBuffer;
 import java.nio.ByteOrder;
 import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
+import org.apache.avro.Schema;
```

Comment:
I'd prefer not passing in the Avro schema. I think that the behavior should be that the `AvroSchemaWithTypeVisitor` visits each union branch and produces a `ValueReader`. Then the visitor implementation should create the index map and pass it into the reader. Not passing the schema in should keep the reader simple.